src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+    componentMapping.r = asset::ICPUImageView::SComponentMapping::ES_ZERO;
+    componentMapping.g = asset::ICPUImageView::SComponentMapping::ES_ZERO;
+    componentMapping.b = asset::ICPUImageView::SComponentMapping::ES_ZERO;
+    componentMapping.a = asset::ICPUImageView::SComponentMapping::ES_ONE;
+    auto srcChannelCount = asset::getFormatChannelCount(srcImageFormat);
+    for (uint32_t c = 0u; c < srcChannelCount; c++)
+        componentMapping[c] = asset::ICPUImageView::SComponentMapping::ES_IDENTITY;
+
+    uint32_t eachBlockNeededMemory  = imageExtentBlockStridesInBytes[0];  // = blockByteSize
+    uint32_t eachRowNeededMemory    = imageExtentBlockStridesInBytes[1];  // = blockByteSize * imageExtentInBlocks.x
+    uint32_t eachSliceNeededMemory  = imageExtentBlockStridesInBytes[2];  // = blockByteSize * imageExtentInBlocks.x * imageExtentInBlocks.y
+    uint32_t eachLayerNeededMemory  = imageExtentBlockStridesInBytes[3];  // = blockByteSize * imageExtentInBlocks.x * imageExtentInBlocks.y * imageExtentInBlocks.z
+
+    // There is remaining layers in region that needs copying
+    uint32_t uploadableArrayLayers = availableMemory / eachLayerNeededMemory;
+    uint32_t remainingLayers = mainRegion.imageSubresource.layerCount - currentLayerInRegion;
+    uploadableArrayLayers = core::min(uploadableArrayLayers, remainingLayers);
+    // A: There is remaining layers left in region -> Copy Slices (Depths)
+    uint32_t uploadableSlices = availableMemory / eachSliceNeededMemory;
+    uint32_t remainingSlices = imageExtentInBlocks.z - currentSliceInLayer;
+    uploadableSlices = core::min(uploadableSlices, remainingSlices);
+    if(uploadableSlices > 0 && minImageTransferGranularity.depth > 1u && (imageOffsetInBlocks.z + currentSliceInLayer + uploadableSlices) < subresourceSizeInBlocks.z)
+        uploadableSlices = core::alignDown(uploadableSlices, minImageTransferGranularity.depth);
+    // B: There is remaining slices left in layer -> Copy Rows
+    uint32_t uploadableRows = availableMemory / eachRowNeededMemory;
+    uint32_t remainingRows = imageExtentInBlocks.y - currentRowInSlice;
+    uploadableRows = core::min(uploadableRows, remainingRows);
+    if(uploadableRows > 0 && minImageTransferGranularity.height > 1u && (imageOffsetInBlocks.y + currentRowInSlice + uploadableRows) < subresourceSizeInBlocks.y)
+        uploadableRows = core::alignDown(uploadableRows, minImageTransferGranularity.height);
+    // C: There is remaining slices left in layer -> Copy Blocks
+    uint32_t uploadableBlocks = availableMemory / eachBlockNeededMemory;
+    uint32_t remainingBlocks = imageExtentInBlocks.x - currentBlockInRow;
+    uploadableBlocks = core::min(uploadableBlocks, remainingBlocks);
+    if(uploadableBlocks > 0 && minImageTransferGranularity.width > 1u && (imageOffsetInBlocks.x + currentBlockInRow + uploadableBlocks) < subresourceSizeInBlocks.x)
+        uploadableBlocks = core::alignDown(uploadableBlocks, minImageTransferGranularity.width);
```

Comment:
You don't need all these `uploadable`

Just be symmetric, and go in  this order:
1. if `currentBlockInRow<imageExtentInBlocks.x` finish any previously incomplete row (upload blocks) 
2. if `currentRowInSlice<imageExtentInBlocks.y` finish any previously incomplete slice (upload rows)
3. if `currentSliceInLayer<imageExtentInBlocks.z` finish any previously incomplete layer (upload slices)
4. upload as many whole layers as you can
5. upload as many slices in the "current" layer as you can
6. upload as many rows in the "current" slice as you can
7. upload as many blocks in the "current" row as you can

Reply:
We only return 1 `regionToCopyNext` this way you'd need to return multiple and it wouldn't be an iterator anymore.
unless you mean the order of the `if`s  which is debatable but you'd need to compute these values at some point, I can change their orders or move them into the `if` clauses if it seems more plausible to you

Reply:
> We only return 1 regionToCopyNext this way you'd need to return multiple and it wouldn't be an iterator anymore.

ok missed that, in that case you only need 1-4 and for them to be exclusive to each other
```
if (currentBlockInRow<imageExtentInBlocks.x)
{
   // case 1
}
else if (currentRowInSlice<imageExtentInBlocks.y)
{
   // case 1
}
else if (currentSliceInLayer<imageExtentInBlocks.z)
{
   // case 1
}
else if (currentLayer<imageExtentInBlocks.w)
{
   // case 4
}
else
   return false; // no region produces
```

Reply:
this won't work either, because the first time when everything is `0` it will try to fill blocks when it may be able to transfer whole regions  (+ they will reset to 0 when hitting the extents)
I don't see anything wrong or unintuative about the current way the `if`s are structured,

note that you may try to fill a row because of 2 different reasons:
1. not enough memory (hence the uploadableXXX)
2. currentBlockInRow != 0

We can do something like this:
```c
    if(currentBlockInRow != 0 || (notEnoughMemoryForCompleteRow && uploadableBlocks > 0))
    {
        // Try Fill row
    }
    else if(currentRowInSlice != 0 || (notEnoughMemoryForCompleteSlice && uploadableRows > 0))
    {
        // Try Fill slice
    }
    else if(currentSliceInLayer != 0 || (notEnoughMemoryForCompleteLayers && uploadableSlices > 0))
    {
        // Try Fill layer
    }
    else if(currentLayerInRegion != 0 || (uploadableArrayLayers > 0))
    {
        // Try Fill region
    }
```

** didn't include `canTransferMipLevelsPartially` for simplicity. **

which is basically the way it currently is but reversed, so let's not touch it?

Reply:
> I don't see anything wrong or unintuative about the current way the ifs are structured,

ok lets not touch the ifs for now, but lets tackle everything else.